rust/src/response_broker.rs

```diff
@@ -0,0 +1,89 @@
+use futures_util::FutureExt;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::future::Future;
+use std::sync::Mutex;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::error::RecvError;
+use tokio::sync::oneshot::Sender;
+use tracing::{error, debug};
+
+#[derive(Debug)]
+pub(crate) struct ResponseBroker<T: Clone + Debug, E: std::error::Error + Clone + From<RecvError>> {
+    responses: Mutex<HashMap<String, Sender<Result<T, E>>>>,
+}
+
+impl<T: Clone + Debug, E: From<RecvError> + Clone + std::error::Error> ResponseBroker<T, E> {
+    pub fn new() -> Self {
+        Self {
+            responses: Mutex::new(HashMap::new()),
+        }
+    }
+
+    pub fn get_response(&self, request_id: String) -> impl Future<Output = Result<T, E>> {
+        let receiver = {
+            let mut requests = self.responses.lock().unwrap_or_else(|e| e.into_inner());
+            let (sender, receiver) = oneshot::channel();
+            requests.insert(request_id, sender);
+            receiver
+        };
+        receiver.map(|r| r?)
+    }
+
+    pub fn set_response(&self, request_id: &str, response: Result<T, E>) {
+        let mut requests = self.responses.lock().unwrap_or_else(|e| e.into_inner());
+        if let Some(r) = requests.remove(request_id) {
+            if r.send(response).is_err() {
+                debug!("Receiver for {request_id} has been dropped");
```

Comment:
IHMO this should be `error!` not `debug!` as file was not scanned successfully and no verdict is available.

Reply:
This is debug, because it can happen during normal operation - a caller can go away due to timeout.